packages/tools/src/utilities/math/polyline/combinePolyline.ts

```diff
@@ -0,0 +1,327 @@
+      }
+
+      targetPolylinePoints.push(targetEdgePoint);
+      sourceIntersectionPoints.push(sourceEdgePoint);
+
+      // Switches from "exiting" to "entering" and vice-versa
+      intersectionPointDirection *= -1;
+    });
+  }
+
+  // Store all vertices and intersections for source contour
+  for (let i = 0, len = sourcePolyline.length; i < len; i++) {
+    const lineSegmentId: number = i;
+    const p1 = sourcePolyline[i];
+    const vertexPoint: PolylinePoint = {
+      type: PolylinePointType.Vertex,
+      coordinates: p1,
+      visited: false,
+      next: null,
+    };
+
+    sourcePolylinePoints.push(vertexPoint);
+
+    const sourceIntersectionPoints =
+      sourceIntersectionsCache.get(lineSegmentId);
+
+    if (!sourceIntersectionPoints?.length) {
+      continue;
+    }
+
+    // Calculate the distance between each intersection point to the start point
+    // of the line segment, sort them by distance and return a sorted array that
+    // contains all intersection points.
+    sourceIntersectionPoints
+      .map((intersectionPoint) => ({
+        intersectionPoint,
+        lineSegStartDistSquared: math.point.distanceToPointSquared(
+          p1,
+          intersectionPoint.coordinates
+        ),
+      }))
+      .sort(
+        (left, right) =>
+          left.lineSegStartDistSquared - right.lineSegStartDistSquared
+      )
+      .map(({ intersectionPoint }) => intersectionPoint)
+      .forEach((intersectionPoint) =>
+        sourcePolylinePoints.push(intersectionPoint)
+      );
+  }
+
+  ensuresNextPointers(targetPolylinePoints);
+  ensuresNextPointers(sourcePolylinePoints);
+
+  return { targetPolylinePoints, sourcePolylinePoints };
+}
+
+function getUnvisitedOutsidePoint(polylinePoints: PolylinePoint[]) {
+  for (let i = 0, len = polylinePoints.length; i < len; i++) {
+    const point = polylinePoints[i];
+
+    if (!point.visited && point.position === PolylinePointPosition.Outside) {
+      return point;
+    }
+  }
+}
+
+/**
+ * Merge two planar polylines (2D)
+ */
+function mergePolylines(
+  targetPolyline: Types.Point2[],
+  sourcePolyline: Types.Point2[]
+) {
+  const targetNormal = getNormal2(targetPolyline);
+  const sourceNormal = getNormal2(sourcePolyline);
+  const dotNormals = vec3.dot(sourceNormal, targetNormal);
+
+  // Both polylines need to be CW or CCW to be merged and one of them needs to
+  // be reversed if theirs orientation are not the same
+  if (!glMatrix.equals(1, dotNormals)) {
+    sourcePolyline = sourcePolyline.reverse();
+  }
+
+  const { targetPolylinePoints } = getSourceAndTargetPointsList(
+    targetPolyline,
+    sourcePolyline
+  );
+  const startPoint: PolylinePoint =
+    getUnvisitedOutsidePoint(targetPolylinePoints);
+
+  // Source polyline contains target polyline
+  if (!startPoint) {
+    return targetPolyline.slice();
+  }
+
+  const mergedPolyline = [startPoint.coordinates];
+  let currentPoint = startPoint.next;
+
+  while (currentPoint !== startPoint) {
```

Comment:
This is most likely the infinite loop you showed me

Reply:
Yes but the problem is related to `getSourceAndTargetPointsList` that is responsible for creating the linked list.